enterprise/ssl-impl/src/main/java/io/crate/protocols/ssl/FilesWatcher.java

```diff
@@ -0,0 +1,108 @@
+ *
+ * To use this file, Crate.io must have given you permission to enable and
+ * use such Enterprise Features and you must have a valid Enterprise or
+ * Subscription Agreement with Crate.io.  If you enable or use the Enterprise
+ * Features, you represent and warrant that you have a valid Enterprise or
+ * Subscription Agreement with Crate.io.  Your use of the Enterprise Features
+ * if governed by the terms and conditions of your Enterprise or Subscription
+ * Agreement with Crate.io.
+ */
+
+package io.crate.protocols.ssl;
+
+import java.io.IOException;
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static java.nio.file.StandardWatchEventKinds.OVERFLOW;
+
+/**
+ * The file watcher checks for changes on registered files
+ * and notifies its listeners accordingly.
+ */
+class FilesWatcher implements Runnable {
+
+    private final WatchService watchService;
+    private final Map<WatchKey, List<FileWatcher>> watchers = new HashMap<>();
+
+    FilesWatcher() throws IOException {
+        this.watchService = FileSystems.getDefault().newWatchService();
+    }
+
+    void register(FileWatcher watcher) throws IOException {
+        WatchKey key = watcher.registrablePath().register(watchService, watcher.events());
+        if (watchers.containsKey(key)) {
+            watchers.get(key).add(watcher);
+        } else {
+            watchers.put(key, new ArrayList<>(Collections.singletonList(watcher)));
+        }
+    }
+
+    @Override
+    public void run() {
+        poll();
+    }
+
+    /**
+     * Polls for watch events on the registered file watchers
+     * and notifies their listeners if events occur.
+     */
+    private void poll() {
+        // break the loop if no watch keys are present in the queue
+        // or watch key events were consumed
+        while (true) {
+            WatchKey key = watchService.poll();
+            if (key == null || !watchers.containsKey(key)) {
+                break;
+            }
+
+            for (WatchEvent<?> event : key.pollEvents()) {
+                WatchEvent.Kind<?> kind = event.kind();
+                if (OVERFLOW == kind) {
+                    continue;
+                }
+                //noinspection unchecked
+                notifyFileWatchers(watchers.get(key), (WatchEvent<Path>) event);
+
+                boolean valid = key.reset();
+                if (!valid) {
+                    break;
+                }
+            }
+        }
+    }
+
+    private void notifyFileWatchers(List<FileWatcher> watchers, WatchEvent<Path> watchEvent) {
+        Path modifiedPath = watchEvent.context();
+        for (var watcher : watchers) {
+            if (watcher.path().toFile().getName().equals(modifiedPath.toFile().getName())) {
```

Comment:
Can the `equals` be done on the `Path` instances directly to avoid the `toFile` conversions?

Reply:
Comparing on `getFileName()` would work then I guess?

```
jshell> Paths.get("/tmp/x").toAbsolutePath().getFileName();
$3 ==> x

jshell> Paths.get("/tmp/x").getFileName();
$4 ==> x
```